### DISCUSSION TITLE
ErrorResponse 추상화

### DISCUSSION THREAD

**Discussion by NamKey** [f-lab-edu/yousinsa#15] [Ideas]
- Exception에 대한 내용을 클라이언트로 반환해야 할 때 보내는 형식에 대한 추상화가 필요합니다.
- 아래 첨부되어 있는 링크 방식대로 하는 것도 좋지만 우리 프로젝트내에서 사용하는 것인만큼 어떻게 사용하면 더 클라이언트에게 명확하게 응답을 보내줄 수 있는지에 대한 부분이 논의되면 좋겠습니다.

[Spring Exception 전략](https://cheese10yun.github.io/spring-guide-exception/ )

**Comment by maketheworldwise** [comment]
첨부해주신 블로그의 내용대로 진행해도 괜찮다고 생각합니다. 

개인적으로는 결과를 반환할 때나, 예외 처리를 할 때나 - 리스폰스의 경우 일관된 형식으로 진행되면 좋을 것 같습니다. 

```json
{
    "status" : "success",
    "code" : 2000,
    "result" : "결과"
}
```

정확히 어떤 부분에서 고민이 되셨나요??

**Reply by NamKey** [reply]
현재는 Exception 관련된 부분만 논의하면 (Response 관련된 부분 제외)
해당 블로그 내용도 좋지만 이렇게 진행했었을 때 우리한테는 적합한가를 고민하고 싶었습니다.

예를 들면 다음 코드에서
- message, status, errors, code에는 어떤 내용이 들어갈 것인가?
- 어느 정도의 정보까지 나타낼 것인가? Field의 값을 잘못 넣었다. Field의 형식에 맞지 않다.
- 더 넣거나 제외할 정보는 없나?

입니다.

``` java
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class ErrorResponse {

    private String message;
    private int status;
    private List<FieldError> errors;
    private String code;
    ...

    @Getter
    @NoArgsConstructor(access = AccessLevel.PROTECTED)
    public static class FieldError {
        private String field;
        private String value;
        private String reason;
        ...
    }
}
```

**Comment by maketheworldwise** [comment]
이렇게 세 가지 항목으로만 응답을 만들어내는 것은 어떨까요?

- status : 성공/실패 여부
- code : Http Status 코드와는 다른 서버 자체 코드 (ex. 성공은 2000번대, 실패는 3000번대 ...)
- result : 결과 / 에러 메시지

저는 에러와 리스폰스를 별도로 구분지어 생각하지 않고 - 에러든, 요청에 대한 결과든, 모든 응답은 동일한 구성으로 진행되어야 개발할 때 편하고 좋다고 생각합니다. 매번 요청에 따라 다른 형식의 응답이 반환되도록 구성한다면 오히려 개발할 때 더 복잡해지는 것 같은 느낌이 들어서 일관성을 유지하는 방법이 좋지 않을까 싶습니다.
